### PR TITLE
[BREAKING] Wrong argument count in constructor call

### PR DESCRIPTION
**Pull request by erak**

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -551,31 +551,18 @@ void TypeChecker::endVisit(InheritanceSpecifier const& _inheritance)
...
+				toString(arguments->size()) +
+				" arguments given but expected " +
+				toString(parameterTypes.size()) +
+				"."
+			);
+			if (arguments->size() != 0)
 				return;
-			}
 		}
 		for (size_t i = 0; i < arguments->size(); ++i)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please use `min(argumentsize, parametersieze)` and remove the return above.

File: Changelog.md [hunk 2]
```diff
@@ -28,6 +28,7 @@ Breaking Changes:
  * Optimizer: Remove the no-op ``PUSH1 0 NOT AND`` sequence.
  * Parser: Disallow trailing dots that are not followed by a number.
  * Parser: Remove ``constant`` as function state mutability modifer.
+ * Type Checker: Disallow calling constructor with wrong argument count. This was already the case in the experimental 0.5.0 mode.
```
Comment (chriseth) [inline Changelog.md]: `Disallow empty parentheses for constructor call if it requires arguments.`

File: test/libsolidity/syntaxTests/inheritance/base_arguments_empty_parentheses.sol [hunk 3]
```diff
@@ -1,7 +1,13 @@
 contract Base {
   constructor(uint) public {}
 }
+contract Base2 {
+  constructor(uint, uint) public {}
+}
 contract Derived is Base(2) { }
 contract Derived2 is Base(), Derived() { }
+contract Derived3 is Base2(2,2) { }
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inheritance/base_arguments_empty_parentheses.sol]: I don't think these changes to the test are needed.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -551,33 +551,18 @@ void TypeChecker::endVisit(InheritanceSpecifier const& _inheritance)
...
-					toString(parameterTypes.size()) +
-					"."
-				);
-				return;
-			}
+			m_errorReporter.typeError(
+				_inheritance.location(),
+				"Wrong argument count for constructor call: " +
+				toString(arguments->size()) +
+				" arguments given but expected " +
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:557]: Sorry, forgot about that in the last review: We should also add something like `Remove parenthesis if you do not want to provide arguments here.`

Comment (erak) [reply libsolidity/analysis/TypeChecker.cpp:557]: I think the suggestion makes sense here, even though I was a bit undecided about it been displayed in this scenario: https://github.com/ethereum/solidity/pull/4454/commits/be29fd5b3c677b9492be8cb2f31ba8d27216e3cf#diff-c03f6d6e2ee8783cb11d4cf3034a52b6

**Review by chriseth — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -548,31 +548,18 @@ void TypeChecker::endVisit(InheritanceSpecifier const& _inheritance)
...
+				toString(parameterTypes.size()) +
+				"."
+			);
+			if (arguments->size() != 0)
 				return;
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: I'm not clear about this. Why would we stop if the argument count is non-zero, but progress otherwise?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Can also be replaced by a `min(. , .)` call in the for-loop below, yes.

**Pull request merged**